Map View and `GeoPoints` archetype

tfoldi opened the pull request:
### What

Map Space View for GNSS/GPS data logged through new GPS Coordinates archetype.    

https://github.com/rerun-io/rerun/assets/82426/86b13ba3-97af-47d2-9422-7052d7064265


- Added a new `re_space_view_map` crate for visualizing `Points3D` with gps coordinates. 
  - `map_visualizer_system.rs` implements the `VisualizerSystem` trait. It has only one meaningful function, that takes the `Positions3D` components and turns into `vec<MapEntry>` array along with optional `Color` and `Radius` components. This `vec<MapEntry>` is the only field of the `MapVisualizerSystem`. 
  - `map_space_view.rs` is where the magic happens. It uses `walkers` slippy map implementation to show an OpenStreetMap or Mapbox map. On the `selection_ui ` you can change the map providers. To view Mapbox maps, you need to set mapbox tokens (you can get free tokens from Mapbox's site) and pass it via the env variable (`MAPBOX_ACCESS_TOKEN`) or via the blueprint
  - `map_windows.rs` is a small helper to show the zoom controls and acknowledgment on the UIs. 
- For blueprints I made `MapProvider` component (enum with providers+styles), and a `MapOptions` archetype. 

The logging of data is performed as follows:

```rust
//! Log some GPS coordinates

fn main() -> Result<(), Box<dyn std::error::Error>> {
    let rec = rerun::RecordingStreamBuilder::new("rerun_example_gps_coordinates").spawn()?;

    rec.log(
        "points",
        &rerun::Points3D::new([(47.6344, 19.1397, 0.0), (47.6344, 19.1399, 1.0)]),
    )?;

    Ok(())
}
```

The blueprint is defined as follows:

```python
"""Use a blueprint to show a map."""

import rerun as rr
import rerun.blueprint as rrb

rr.init("rerun_example_gps_coordinates", spawn=True)

rr.log("points", Points3D([[47.6344, 19.1397, 0], [47.6334, 19.1399, 1]]))

# Create a map view to display the chart.
blueprint = rrb.Blueprint(
    rrb.MapView(
        origin="points",
        name="MapView",
        map_options=rrb.archetypes.MapOptions(provider=rrb.components.MapProvider.MapboxStreets),
    ),
    collapse_panels=True,
)

rr.send_blueprint(blueprint)
```

The PR is not perfect, but the view is usable.  

Please review the changes and evaluate their potential utility for your needs. If you don't like, prefer not to merge just close it - no hard feelings. 

### To be discussed

 * Check with you guys to see if the name is right at all. Maybe instead of GPS we can go with something better
 * Position3D is limited to `f32` instead of `f64`. This makes it less precise than `/NavSatFix` 


### Checklist
* [x] I have read and agree to [Contributor Guide](https://github.com/rerun-io/rerun/blob/main/CONTRIBUTING.md) and the [Code of Conduct](https://github.com/rerun-io/rerun/blob/main/CODE_OF_CONDUCT.md)
* [x] I've included a screenshot or gif (if applicable)
* [x] I have tested the web demo (if applicable):
  * Using my own lame demo app which is not part of the PR. 
* [x] The PR title and labels are set such as to maximize their usefulness for the next release's CHANGELOG


- [PR Build Summary](https://build.rerun.io/pr/6561)
- [Recent benchmark results](https://build.rerun.io/graphs/crates.html)
- [Wasm size tracking](https://build.rerun.io/graphs/sizes.html)

To run all checks from `main`, comment on the PR with `@rerun-bot full-check`.

k-bx commented:
Hey, sorry for a noob question, but how do I check this out?

```
ubuntu@echo-gpu:~/workspace/rerun-tfoldi$ pixi run rerun-web-release
 Pixi task (rerun-build-web-release in default): rustup target add wasm32-unknown-unknown && cargo run -p re_dev_tools -- build-web-viewer --release -g
info: component 'rust-std' for target 'wasm32-unknown-unknown' is up to date
    Finished dev [optimized + debuginfo] target(s) in 0.10s
     Running `target/debug/re_dev_tools build-web-viewer --release -g`
Building web viewer…

Compiling Rust to wasm in /home/ubuntu/workspace/rerun-tfoldi/target_wasm…
/home/ubuntu/workspace/rerun-tfoldi> CARGO_ENCODED_RUSTFLAGS="--cfg=web_sys_unstable_apis" RUSTFLAGS="--cfg=web_sys_unstable_apis" "cargo" "build" "--quiet" "--package" "re_viewer" "--lib" "--target" "wasm32-unknown-unknown" "--target-dir" "/home/ubuntu/workspace/rerun-tfoldi/target_wasm" "--no-default-features" "--features=analytics" "--release"
error: package `bitstream-io v2.4.2` cannot be built because it requires rustc 1.79 or newer, while the currently active rustc version is 1.76.0
Either upgrade to rustc 1.79 or newer, or use
cargo update bitstream-io@2.4.2 --precise ver
where `ver` is the latest version of `bitstream-io` supporting rustc 1.76.0
Error: Failed to build Wasm
```

tfoldi commented:
> Either upgrade to rustc 1.79 or newer, or use
> cargo update bitstream-io@2.4.2 --precise ver
> where `ver` is the latest version of `bitstream-io` supporting rustc 1.76.0
> Error: Failed to build Wasm

New rerun needs rust version 1.79 while you have only 1.76. Update your rust toolchain and try it again.

k-bx commented:
Ah, needed to update the `rust-toolchain` file. All good, compiled fine!

tfoldi commented:
Few changes we discussed:

 - Remove `GpsCoordinate` archetype 
 - Add `ViewCoordinates.LAT_LON_ALT` enum value to indicate that a `Position3D` represents a lat/lon/alt values
 - Use `ViewCoordinates` in view heuristics to select the right heuristics for view
 - due to `walkers`' `reqwest` dependency it should be an optional feature

jleibs commented:
>     * Add `ViewCoordinates.LAT_LON_ALT` enum value to indicate that a `Position3D` represents lat/lon/alt values 
>     * Use `ViewCoordinates` in view heuristics to select the right heuristics for view

I talked this one over with some folks and the desire is to keep ViewCoordinates separate and somewhat dedicated to interpretation of camera-orientation, as distinct from data semantics.

I'm going to put together a proposal for a new archetype, probably named something like CRS (CoordinateReferenceSystem). We can start with this somewhat dedicated to the mapping use-cases, but I think it could be helpful for other things like CAD import where it might be useful to carry context like unit-sizes.

jleibs commented:
Here's my proposal for adding a very simple CRS type as an indicator: https://github.com/rerun-io/rerun/issues/6601

tfoldi commented:
The new `image` crate depends on a newer version of Rust by default unless you disable it by setting `default-features = false`. Since `walkers` enables default features, this is causing the Rust build to fail on CI. I will look into how this can be addressed in the original repository.

The current cargo tree:
```
    ├── re_data_source v0.17.0-alpha.4 (/Users/.../rerun/crates/re_data_source)
    │   ├── re_data_loader v0.17.0-alpha.4 (/Users/.../rerun/crates/re_data_loader)
    │   │   ├── image v0.25.1
    │   │   │   ├── ravif v0.11.5
    │   │   │   │   ├── rav1e v0.7.1
    │   │   │   │   │   ├── bitstream-io v2.4.2  <--- this requires 1.79
```

In `image`'s `Cargo.toml`:
```
ravif = { version = "0.11.2", default-features = false, optional = true }
```

tfoldi commented:
- Removed `GpsCoordinates` in favor of pure `Points3D`. 
 - Removed default features from `image` in `walkers` repo so now the code works with older `rustc`.  
 - Formatted/linted/removed all gps position related stuff 

What is left from what we discussed:
1. add feature flags to make the map view optional
2. add `ZoomLevel` and `Secret` components for the map options archetype
   - For `zoom level` I plan to add a component based on float. I will call it as zoom level, but maybe map zoom level would be more specific (but less reusable). 
   - `Secret` component should be reusable and displayed as password field in `selection_ui`

For 1, I'm open to your suggestions. Should the feature be enabled or disabled by default? Either way, how should the CI/CD integration work? Should we test all features by default or test combinations of features?

patrickelectric commented:
@tfoldi nice work! Excited to see it merged :)

tfoldi commented:
Extended the nuScenes example with MapView to have at least one example using the new view:

https://github.com/rerun-io/rerun/assets/82426/8b329788-a928-4f5a-b228-514642104203

journaux commented:
would this support plotting a scatter of points across the map (instead of a single point/track) ?

tfoldi commented:
> would this support plotting a scatter of points across the map (instead of a single point/track) ? from the PR seems yes a la `rr.log("points", Points3D([[47.6344, 19.1397, 0], [47.6334, 19.1399, 1]]))` however the visual confused me (centers on a single point)

yes, it does support multiple points

journaux commented:
good to know ! curious if team/egui has evaluated integrating cesium

-> its rendering pipeline for large set of N points may be faster
-> it has 3d terrain capability

non-trivial so aware might also be out of scope

k-bx commented:
Hey, still very keen to see this. What's the current blocker?

samorr commented:
The same from my side, will it be continued? 
Is there any plan to make GPS coordinates using `float64`? With 32 bits it's very 'rough'

tfoldi commented:
> The same from my side, will it be continued?

I’m open to continuing this, but it's largely dependent on the Rerun team's priorities. As you may have seen from the PR history, I tried to keep it up-to-date for a few versions but eventually paused due to the complexity involved. There are many moving parts in Rerun, and with so much auto-generated code and changing internals, maintaining consistency across view-related PRs isn’t straightforward.

I received two key comments from @jleibs:
1. #6601: Suggestion to introduce a new `CoordinateReferenceSystem` type as type indicator. Personally, I’m not sure this is necessary on day one. We could potentially move forward by logging `Position3D` types and setting the view to `Map` in the blueprint. 
2. The `walker` egui component brings in the `reqwest` crate, so making it an optional feature might be a cleaner approach. 


On June 22nd, I asked a few questions regarding these points, but since I didn’t receive any responses, so I paused further updates. That said, I’m happy to rebase the PR to the latest main branch if there’s a realistic chance it could be merged.

abey79 commented:
@tfoldi First, thanks again for your contribution, and for putting up with our slow responses so far. With 0.19 out and the huge push on the dataframe and video fronts, we are now ready to focus on other areas—and this one is high up our list. Starting on Monday, I'll make it my priority to help you land this PR.

Would you be willing to rebase it to latest main? My plan would then be to dive in, give it a thorough review, and help (including hands-on) to land it sooner than later.

As for the "feature flag": yes we want one, that is disabled by default (at least initially). This will give us more confidence to land stuff before they are 100% polished. To be clear, by "feature flag", we don't necessarily mean a Cargo-style feature, but a global application setting which dynamically enable/disable the map view in the viewer. That's often how we signify to users that a particular view is "in beta". (For illustration, I recently removed the feature flag we had on the dataframe view in [this PR](https://github.com/rerun-io/rerun/pull/7663)).

abey79 commented on crates/store/re_types/definitions/rerun/archetypes/points3d.fbs:
```diff
@@ -11,7 +11,7 @@ namespace rerun.archetypes;
 table Points3D (
```
As discussed, this is probably no the right archetype to initially support. Either Point2D or some custom type.

abey79 commented on crates/viewer/re_space_view_map/src/map_space_view.rs:
```diff
@@ -0,0 +1,334 @@
+    tiles: Option<HttpTiles>,
+    map_memory: MapMemory,
+    selected_provider: MapProvider,
+    access_token: String,
```
I think `selected_provider` and `access_token` are not really necessary here. We usually try to keep view states as lean as possible. My suggestion:
- remove these two fields
- make `map_memory` an `Option`
- update `map_memory` from `MapViewClass::ui()`

Also, I'd either inline `ensure_and_get_mut_refs` or somehow move it to `MapViewClass`.

abey79 replied:
You're right. Then let's just add a comment above `selected_provider` to indicate that it is used for invalidation.

abey79 commented on crates/viewer/re_space_view_map/src/map_space_view.rs:
```diff
@@ -0,0 +1,334 @@
+            );
+
+            map_widget.double_clicked().then(|| {
+                map_memory.follow_my_position();
```
Our usual semantics for double clicking is zoom to the "bounding box", e.g. whatever area that contains the entire view content.

I haven't looked into it yet. If it is involved, we can keep that for a follow-up PR though.

tfoldi replied:
double click reset the view (re-enable follow mode). We can change the zoom level as well, but if you have only point finding what zoom level the user expects is tricky imu

abey79 reviewed (commented):
This generally looks in excellent shape. Thanks for this contribution! I'm not yet completely done with my review, here are already some comments.

abey79 commented on crates/store/re_types/definitions/rerun/blueprint/archetypes/map_options.fbs:
```diff
@@ -0,0 +1,21 @@
+    // --- Optional ---
+
+    /// Zoom level for the map. The default is 16.
+    zoom: rerun.blueprint.components.ZoomLevel ("attr.rerun.component_optional", order: 2000);
```
In addition to zoom level, it would be extremely valuable to store in the blueprint the scroll position (coordinate of center point?). In particular, that would make it settable from code.

If that's too involved, we can keep it for a follow-up PR.

tfoldi replied:
absolutely possible, but the map center can be more than a position:

 - it can be in follow mode, which is useful if you have one or interconnected points. here the first logged point is always in the center (like in the embedded video on the PR description)
 - we can fix the map at a certain lat, long position

what archetypes should I use to store this? in the lib if the map center is `None` then it is in the following mode; otherwise, it is fixed. but I assume this approach is not idiomatic in rerun

tfoldi replied:
According to your last comment, the `center_position` should be `DVec2D`, and `follow_mode` should be a new enum or a boolean with a well-named entry to keep the number of components down.

abey79 replied:
My gut feeling would be to minimise the feature set for this PR and skip any sort of "auto-scroll" mode. The MVP would be:
- `CenterPosition` view property (datatype: `DVec2D`) to store the center location along with the zoom level
- Best effort implementation for the fallback provider of `CenterPosition` based on whatever data is logged

In the future, we indeed need to devise a set of more advanced auto-scroll/follow mode, but I think the hindsight we'll gain from landing this PR will help designing that properly. (Ideally, it'd be somewhat consistent with the 2D view too.)

abey79 commented on crates/viewer/re_space_view_map/src/map_space_view.rs:
```diff
@@ -0,0 +1,334 @@
+            map_windows::acknowledge(ui, &window_id, &map_pos, tiles.attribution());
+
+            // update blueprint if zoom level changed from ui
+            if map_memory.zoom() != *zoom_level {
```
see my other note, would be great to store the map position in the blueprint as well

tfoldi commented on crates/viewer/re_viewer_context/src/space_view/mod.rs:
```diff
@@ -61,6 +61,9 @@ pub enum SpaceViewSystemExecutionError {
     DrawDataCreationError(Box<dyn std::error::Error>),
 
+    #[error("Error accessing map view tiles.")]
+    MapTilesError,
```
yes, I can add the underlying HTTP error message

abey79 commented:
I just discussed the data model aspect of this PR with @jleibs and we reached the following decision.

We introduce a new `GeoPoints` archetype as per this outline:

```
table GeoPoints  {
  /// All the 3D positions at which the point cloud shows points.
  latlon: [rerun.components.LatLon];  // datatype: DVec2D

  // --- Recommended ---

  /// Optional radii for the points, effectively turning them into circles.
  radii: [rerun.components.Radius];

  /// Optional colors for the points.
  colors: [rerun.components.Color];

  // --- Optional ---

  /// Optional text labels for the points.
  labels: [rerun.components.Text];

  /// Optional choice of whether the text labels should be shown by default.
  show_labels: rerun.components.ShowLabels;
}
```
This PR will have to introduce the `DVec2D` datatype, identical to the `Vec2D` but with double precision floats.

For the scope of this PR, the recommended/optional components should be included only so long as they are actually supported by the view. If they are not, then it's fine to drop them—we can add them in a subsequent PR.

In the future, we will further add a `Altitude` recommended component to carry the height information. How we define the kind of altitude we support (ground height, ellipsoid, amsl, whatnot) is yet to be defined. We currently lean towards having that setting as a view property, but that's for another day to decide.

**Motivation for this decision**:
- Using `Points2D` is problematic precision-wise.  `f32` means ~1.7m real world accuracy. `GeoPoints` has `f64`.
- Splitting the latlon part from the (future) Altitude component clarifies that these numbers dont have the same units (angles for latlon, likely meters for Altitude).
- More generally, we will try to adhere to the standard GIS data model going forward (in particular the [GeoArrow](https://geoarrow.org/format.html) model). The proposed `GeoPoints` archetype fits the `MultiPoint` case (by virtue of our components always being batch), and it's likely that `MultiLineString` and `MultiPolygon` will eventually follow.
- Having a distinct archetype from `Points2D` also has the advantage to disambiguate things for the heuristics.

tfoldi commented:
> For the scope of this PR, the recommended/optional components should be included only so long as they are actually supported by the view. If they are not, then it's fine to drop them—we can add them in a subsequent PR.

we can support text labels. However, shouldn't you be able to change their fonts, sizes, and colors?

abey79 commented:
> > For the scope of this PR, the recommended/optional components should be included only so long as they are actually supported by the view. If they are not, then it's fine to drop them—we can add them in a subsequent PR.
> 
> we can support text labels. However, shouldn't you be able to change their fonts, sizes, and colors?

No, I don't think this is required for this PR (and the short term future tbh). IWe don't support styling of labels in other views either.

abey79 commented:
Let's make sure we reference https://epsg.io/4326 in the `GeoPoint` archetype and `LatLon` component.

emilk commented on crates/top/rerun-cli/Cargo.toml:
```diff
@@ -45,7 +45,7 @@ default = ["native_viewer", "web_viewer"]
 ## The features we enable when we build the pre-built binaries during our releases.
 ## This may enable features that require extra build tools that not everyone heas.
-release = ["default", "nasm"]
+release = ["default", "nasm", "map_view"]
```
We should also add the new feature to 

https://github.com/grtlr/rerun/blob/1408badf5522eddeae82e2ee069771e3510cf7e3/rerun_py/Cargo.toml#L20-L22

abey79 replied:
On second thought: I don't think so?

That feature only controls whether or not the `re_space_view_map` crate is included, and the corresponding view registered. All the data/blueprint archetypes/components are unconditionally included regardless of the cargo feature.

This solves the dependency problem, but doesn't "hide" everything genospatial. Doing so would require way more work, incl. on codegen, so I'd rather avoid that.

With that being said, `rerun_py/Cargo.toml` doesn't depend at all on 
`re_space_view_map`, `re_viewer`, `rerun` or `rerun_cli` (which are affected by that feature).

Wumpf commented on crates/store/re_types/definitions/rerun/blueprint/components/zoom_level.fbs:
```diff
@@ -0,0 +1,13 @@
+    "attr.rerun.scope": "blueprint",
+    "attr.rust.derive": "Default"
+) {
+    /// Zoom level: 0 being the lowest zoom level (fully zoomed out) and 22 being the highest (fully zoomed in).
```
where do these numbers come from? Are they just implementation details of the walkers library?
If so, I think it would make sense for us to abstract this away and instead use a more normalized range like 0-1. 0 to 22 seems rather arbitrary.

tfoldi replied:
tile servers usually implement 18-23 zoom levels starting from zero. 

Google, Mapbox and Azure have max 22 zoom level (except for a few areas where they have 23)

OpenStreetMap maps also vary, but most support levels only up to 18.

with custom maps you go higher. but on zoom level 24 one pixel is 4.77 mm on the map 

in short, there is no theoretical maximum, but drawing a line at 22 is a reasonable choice

abey79 replied:
Yeah this is a tricky topic. Ideally, the max zoom value would be provided by map provider, and then used by the `ZoomLevel` edit ui. @Wumpf can you devise a mechanism that would allow that?

In any case, we're going to have to take shortcuts here. Cut-off at 19 was a bad idea of mine, it's not enough. I will re-instate the cutoff at 22.0 and turn the warning into a `DEBUG` log or something.

This area will need work in the future, including: https://github.com/podusowski/walkers/issues/198

Wumpf replied:
alright! wasn't aware those zoom levels are "a thing", makes sense then

abey79 replied:
**EDIT**: just realised that currently walkers has a hard-coded limit at 19.0 for zoom level. We'll need to improve on that, but for now I'm gonna hard code a limit at 19.0 as well and this will silence most of the log noise (which is in `DEBUG` now anyways).

Wumpf commented on crates/store/re_types/src/components/lat_lon_ext.rs:
```diff
@@ -0,0 +1,47 @@
+// ---
+
+impl LatLon {
+    /// The origin.
+    pub const ZERO: Self = Self::new(0.0, 0.0);
```
That doesn't really make sense for lat/lon. That's just a point in the ocean at the equator

Wumpf commented on crates/store/re_types/src/components/lat_lon_ext.rs:
```diff
@@ -0,0 +1,47 @@
+}
+
+#[cfg(feature = "mint")]
+impl From<LatLon> for mint::Point2<f64> {
```
I don't think these mint conversions make sense. Mint's types are about cartesion type

Wumpf commented on crates/viewer/re_component_ui/src/datatype_uis/float_drag.rs:
```diff
@@ -76,3 +86,17 @@ fn edit_f32_zero_to_one_raw(ui: &mut egui::Ui, value: &mut MaybeMutRef<'_, f32>)
     }
 }
+
+/// Editor for a [`re_types::blueprint::components::ZoomLevel`] value from zero to 19.0.
```
Why 0 to 19, didn't it say on zoom level it was 0 to 22?
If we decide to have this arbitrary range after all, then it should be encoded in constants

abey79 replied:
> Why 0 to 19, didn't it say on zoom level it was 0 to 22?

See other comment. Vast topic 😅

Wumpf commented on crates/store/re_types/definitions/rerun/blueprint/archetypes/map_options.fbs:
```diff
@@ -0,0 +1,18 @@
+    // --- Required ---
+
+    /// Map provider and style to use.
+    provider: rerun.blueprint.components.MapProvider ("attr.rerun.component_required", order: 1000);
```
why is this required?

Wumpf commented on crates/store/re_types/definitions/rerun/blueprint/archetypes/map_options.fbs:
```diff
@@ -0,0 +1,18 @@
+namespace rerun.blueprint.archetypes;
+
+
+/// Configuration for the background of a view.
```
incorrect c&p'ed comment

Wumpf commented on crates/store/re_types/definitions/rerun/blueprint/archetypes/map_options.fbs:
```diff
@@ -0,0 +1,18 @@
+
+
+/// Configuration for the background of a view.
+table MapOptions (
```
I'm not entirely sure if the grouping of "map options" makes sense like this, but I don't have a better idea right now either.
Usually we try to group such that the generic blueprint api & ui becomes meaningful. This now just generates "options" which isn't amazing

abey79 replied:
Right. In the future, we definitely want to expose the map center location or something similar, so that will be grouped with `ZoomLevel`. We can definitely split of the map provider though. Maybe there will be other options in the future falling in the "background" category.

abey79 replied:
Let's go for:
- `area` / `MapArea` / `ZoomLevel` (in the future something like `CenterPosition` TBD)
- `background` / `MapBackground` / `Provider`

Wumpf replied:
Provider seems more fundamental than the others since instead of styling it's more "data content", but otherwise make sense!

Wumpf commented on docs/snippets/all/views/map.py:
```diff
@@ -0,0 +1,20 @@
+    rrb.MapView(
+        origin="points",
+        name="MapView",
+        options=rrb.archetypes.MapOptions(provider=rrb.components.MapProvider.MapboxStreets),
```
mapbox needs `RERUN_MAPBOX_ACCESS_TOKEN` env var, right? So this should probably either not use that or at least document this fact.
Also, we should ideally use all types here, so this should also show how to set the zoom level

abey79 replied:
I'd like to keep that for follow-up PRs. I have these issues that I consider part of the 1st release MVP:
- https://github.com/rerun-io/rerun/issues/7903
- https://github.com/rerun-io/rerun/issues/7849

Wumpf replied:
👍

Wumpf commented on crates/store/re_types/definitions/rerun/blueprint/views/map.fbs:
```diff
@@ -0,0 +1,12 @@
+table MapView (
+    "attr.rerun.view_identifier": "Map"
+) {
+    /// Configures the look and feel of the map.
+    options: rerun.blueprint.archetypes.MapOptions (order: 1000);
```
related to comment on MapOptions - can we think of other future options that would inform a better parameter/field name here? Everything is an option, so calling this option isn't very helpful
(this is a tricky one and I don't have good answers, so if we can't come up with anything better right now let's leave it be)

abey79 replied:
proposal: https://github.com/rerun-io/rerun/pull/6561#discussion_r1816118506

Wumpf commented on crates/store/re_types/definitions/rerun/blueprint/views/map.fbs:
```diff
@@ -0,0 +1,12 @@
+namespace rerun.blueprint.views;
+
+/// A map view.
```
a bit more description of a what map view is capable of would be nice. Keep in mind that this is the main entry point for "these are the view types the viewer has and that's what they do".

Wumpf commented on crates/viewer/re_space_view_map/src/map_space_view.rs:
```diff
@@ -0,0 +1,280 @@
+    fn help_markdown(&self, _egui_ctx: &egui::Context) -> String {
+        "# Map view
+
+Displays a Position3D on a map.
```
outdated comment by now?

Wumpf commented on crates/viewer/re_space_view_map/src/map_space_view.rs:
```diff
@@ -0,0 +1,280 @@
+    }
+
+    fn new_state(&self) -> Box<dyn SpaceViewState> {
+        Box::<MapSpaceViewState>::new(MapSpaceViewState {
+            tiles: None,
+            map_memory: MapMemory::default(),
+            selected_provider: MapProvider::default(),
+        })
```
```suggestion
        Box::<MapSpaceViewState>::new(MapSpaceViewState::default())
```

Wumpf commented on crates/viewer/re_space_view_map/src/map_space_view.rs:
```diff
@@ -0,0 +1,280 @@
+        let zoom_level = blueprint_zoom_level.or(default_zoom_level).unwrap_or(16.0);
+
+        if state.map_memory.set_zoom(zoom_level).is_err() {
+            re_log::warn_once!(
+                "Failed to set zoom level for map. Zoom level should be between zero and 22"
+            );
```
should still set the zoom level to a clamped value I reckon

abey79 replied:
Change that to a debug level signal, because it's not relevant to the user, but we need to be reminded to improve on this situation, which requires upstream work.

abey79 replied:
Also, we now clamp at 22.0, but that can still fail if/when the map provider doesn't have such high res tiles.

abey79 replied:
19.0 it is, see my other, kept not hidden comment.

Wumpf commented on crates/viewer/re_space_view_map/src/visualizers/geo_points.rs:
```diff
@@ -0,0 +1,163 @@
+            let all_colors = results.iter_as(timeline, Color::name());
+            let all_radii = results.iter_as(timeline, Radius::name());
+
+            // default component values
```
nit: convolutes "default" & "fallback". Something I'm very guilty of as well by making fallbacks use defaults by default, but they are distinct things since fallbacks may be a lot more complex than Rust defaults

Wumpf reviewed (changes requested):
Reviewing via request of @abey79

Walkers debug log is a bit spammy, we should add it to `CRATES_AT_INFO_LEVEL` since in the dev environment we default to debug log level otherwise

It's a bit unclear to me which components/archetype should be marked as experimental if any. Blueprint / control of the map is very different from 2D view and we likely want to iterate to get them closer. But given that that likely means iterating on _both_ it's probably moot.

I generally looked over the actual mechanics  of the view only briefly and rough form, docs and whatever else caught my interest ;-) 

Overall pretty amazing piece of work. There's a tiny bit more polish / surprise-reduction I'd like to see even on this first version, see comments :)

abey79 commented on crates/viewer/re_selection_panel/src/visible_time_range_ui.rs:
```diff
@@ -22,6 +22,8 @@ static VISIBLE_HISTORY_SUPPORTED_SPACE_VIEWS: once_cell::sync::Lazy<
         SpatialSpaceView3D::identifier(),
         SpatialSpaceView2D::identifier(),
         TimeSeriesSpaceView::identifier(),
+        // TODO(#7876): replace with `MapSpaceView::identifier()` when we get rid of the cargo feature
```
I'll open a PR for that, sounds easy enough to add `SpaceViewClass::supports_visible_time_range()`

Wumpf reviewed (approved):
🚢

samorr commented:
Hi, maybe this is a little bit premature, but I would very much like to use these functionalities soon. Do you have any estimates on when there's going to be an official release with these changes?

abey79 commented:
> Hi, maybe this is a little bit premature, but I would very much like to use these functionalities soon. Do you have any estimates on when there's going to be an official release with these changes? 

We intend for that to happen relatively soon (couple of weeks ish), but no promises! There should however be at least a development build available sometime next week.

Wumpf commented:
@samorr the latest development build should have this already now! :)
https://github.com/rerun-io/rerun/releases/tag/prerelease

samorr commented:
> @samorr the latest development build should have this already now! :)
> 
> https://github.com/rerun-io/rerun/releases/tag/prerelease

Thanks! I will use it already tomorrow! 😁
